### PR TITLE
Connect chat header component

### PR DESCRIPTION
**Pull request by chrisnojima**
Along the way to making perf better i'm splitting up the major parts of the Conversation component. They really don't interact with each other and we currently split mostly orthogonal properties across them (using the index-hoc). The existing implementation causes a lot of rerender when unrelated parts are changing.

Branches off of this will come for input and sidepanel next

Note: I also added a new debug helper for reselect so you can see when you're cache is being missed

### CODE REVIEW

File: shared/chat/conversation/container.js [hunk 1]
```diff
@@ -52,6 +52,12 @@ class ConversationContainer extends Component<void, Props, State> {
     this.setState({listScrollDownCounter: this.state.listScrollDownCounter + 1})
   }
 
+  // We wrap this so children don't churn when this.props.onBack() changes due to this component churning. Whe this thing does less we can
+  // likely remove this
+  _onBack = () => {
```
Comment (chrisnojima) [inline shared/chat/conversation/container.js]: This component currently churns a bunch which means the caching doesn't work well and therefore the dispatch mapping gets called a bunch which causes this.props.onBack to mutate, and since we pass that as an ownProps to Header that was causing re-render. When this thing is better about its own cache we can remove this.

File: shared/chat/conversation/header/container.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+const getYou = (state: TypedState) => state.config.username || ''
+const getFollowingMap = (state: TypedState) => state.config.following
+const getMetaDataMap = (state: TypedState) => state.chat.get('metaData')
+const getSelectedConversation = (_, selectedConversation: ?Constants.ConversationIDKey) => selectedConversation
+const getSelectedInbox = (state: TypedState, selectedConversation: ?Constants.ConversationIDKey) => (
+  state.chat.get('inbox').find(inbox => inbox.get('conversationIDKey') === selectedConversation)
+)
+
+const getTLF = createSelector(
```
Comment (chrisnojima) [inline shared/chat/conversation/header/container.js]: returning the tlf as its easy to compare. Arrays/Lists/etc won't follow simple equality checks since we're doing filtering and whatnot

File: shared/chat/conversation/header/index.desktop.js [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-import {Box, Icon, Usernames} from '../../common-adapters'
-import {globalStyles, globalColors, globalMargins} from '../../styles'
+import {Box, Icon, Usernames} from '../../../common-adapters'
+import {globalStyles, globalColors, globalMargins} from '../../../styles'
 
-import type {Props} from './header'
+import type {Props} from '.'
 
-const Header = ({muted, onOpenFolder, onShowProfile, onToggleSidePanel, sidePanelOpen, users}: Props) => (
+const ConversationHeader = ({muted, onOpenFolder, onShowProfile, onToggleSidePanel, sidePanelOpen, users}: Props) => (
```
Comment (chrisnojima) [inline shared/chat/conversation/header/index.desktop.js]: renamed this as the banner has a Header also which makes understanding the KBPerf() harder

**Comment by chrisnojima** [comment]
@keybase/react-hackers

File: shared/util/dev.js [hunk 4]
```diff
@@ -8,6 +9,16 @@ const injectReactQueryParams = (url: string): string => {
   return `${url}${url.indexOf('?') === -1 ? '?' : '&'}react_perf`
 }
 
+const debugCreateSelector = createSelectorCreator(
```
Comment (MarcoPolo) [inline shared/util/dev.js]: How do you use this? maybe leave a comment explaining

File: shared/chat/conversation/header/container.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    Constants.usernamesToUserListItem(Constants.participantFilter(List(tlf.split(',')), you).toArray(), you, metaDataMap, followingMap)
+  )
+)
+
+const getMuted = createSelector(
+  [getSelectedInbox],
+  (selectedInbox) => selectedInbox && selectedInbox.get('status') === 'muted',
+)
+
+const mapStateToProps = (state: TypedState, {selectedConversation, sidePanelOpen}: OwnProps) => ({
```
Comment (MarcoPolo) [inline shared/chat/conversation/header/container.js]: selectedConversation can be a selector too, no? but I guess we can do that in another PR

File: shared/chat/conversation/index-hoc.js [hunk 6]
```diff
@@ -106,18 +95,7 @@ const propsHoc = withProps(
       selectedConversation,
     }
 
-    const headerProps: HeaderProps = {
```
Comment (MarcoPolo) [inline shared/chat/conversation/index-hoc.js]: I don't understand why you had to get rid of this:

This is what I mean: https://github.com/keybase/client/commit/16ceed8ecdd2a845a395807283442df976ab29d9

Then you don't need to be copying the same changes across desktop/native. You could control the props passed in once here.

File: shared/chat/conversation/header/container.js (hunk 5 above)
Comment (chrisnojima) [inline shared/chat/conversation/header/container.js]: yes, this was a little tricky as this is actually in the route state and not in our chat store. In my branch off of this I do this so I'll back port this once that lands also

Comment (MarcoPolo) [inline shared/chat/conversation/header/container.js]: We could make a selector that gets the current route's state. Just a thought

Comment (chrisnojima) [inline shared/chat/conversation/header/container.js]: It's already done in the next branches :)

**Comment by keybase-ci-visdiff** [comment]
The commits ff00f423f448...19394ba2918c introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **unlock_folders-Single** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491315151221-ff00f423f448-19394ba2918c-linux/unlock_folders-Single-ff00f423f448.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491315151221-ff00f423f448-19394ba2918c-linux/unlock_folders-Single-19394ba2918c.png)  
   unlock_folders-Single comparison
 * Changed: **Chat_Side_Panel-Muted** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491315151221-ff00f423f448-19394ba2918c-linux/Chat_Side_Panel-Muted-ff00f423f448.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491315151221-ff00f423f448-19394ba2918c-linux/Chat_Side_Panel-Muted-19394ba2918c.png)  
   Chat_Side_Panel-Muted comparison
</summary>

**Comment by chrisnojima** [comment]
Seems like i also fixed the muted visdiff :)
